services/121-service/src/fsp-integrations/account-management/intersolve-visa-account-management/intersolve-visa-account-management.service.ts

```diff
@@ -80,127 +50,135 @@ export class IntersolveVisaAccountManagementService {
+      {
+        referenceId,
+        programId,
+      },
+    );
+
+    await this.intersolveVisaService.hasIntersolveCustomer(registration.id);
+
+    await this.replaceCard({
       referenceId,
       programId,
-      relations: ['programFspConfiguration'],
+      registrationId: registration.id,
+      programFspConfigurationId: registration.programFspConfigurationId,
     });
-    if (
-      !registration.programFspConfigurationId ||
-      registration.programFspConfiguration?.fspName !== Fsps.intersolveVisa
-    ) {
-      throw new HttpException(
-        `This registration is not associated with the Intersolve Visa Fsp.`,
-        HttpStatus.BAD_REQUEST,
-      );
-    }
 
-    const intersolveVisaConfig =
-      await this.programFspConfigurationRepository.getPropertiesByNamesOrThrow({
-        programFspConfigurationId: registration.programFspConfigurationId,
-        names: [
-          FspConfigurationProperties.brandCode,
-          FspConfigurationProperties.coverLetterCode,
-        ],
+    if (userId) {
+      await this.queueMessageService.addMessageJob({
+        registration,
+        messageTemplateKey: ProgramNotificationEnum.reissueVisaCard,
+        messageContentType: MessageContentType.custom,
+        messageProcessType:
+          MessageProcessTypeExtension.smsOrWhatsappTemplateGeneric,
+        userId,
       });
+    }
+  }
 
-    //  TODO: REFACTOR: This 'ugly' code is now also in payments.service.createAndAddIntersolveVisaTransactionJobs. This should be refactored when there's a better way of getting registration data.
-    const intersolveVisaAttributes =
-      FSP_SETTINGS[Fsps.intersolveVisa].attributes;
+  public async replaceCardOnSite({
+    referenceId,
+    programId,
+    tokenCode,
+  }: {
+    referenceId: string;
+    programId: number;
+    tokenCode: string;
+  }): Promise<void> {
+    const registration = await this.registrationsService.getRegistrationOrThrow(
+      {
+        referenceId,
+        programId,
+      },
+    );
 
-    const intersolveVisaAttributeNames = intersolveVisaAttributes.map(
-      (q) => q.name,
+    await this.throwIfCardDistributionByMailEnabled(
+      registration.programFspConfigurationId,
     );
-    const dataFieldNames = [
-      FspAttributes.fullName,
-      FspAttributes.phoneNumber,
-      ...intersolveVisaAttributeNames,
-    ];
-
-    const registrationData =
-      await this.registrationDataScopedRepository.getRegistrationDataArrayByName(
-        registration,
-        dataFieldNames,
-      );
 
-    if (!registrationData || registrationData.length === 0) {
-      throw new HttpException(
-        `No registration data found for referenceId: ${referenceId}`,
-        HttpStatus.NOT_FOUND,
-      );
+    await this.intersolveVisaService.hasIntersolveCustomer(registration.id);
+
+    if (tokenCode) {
+      await this.throwIfCardDoesNotExistOrIsAlreadyLinked(tokenCode);
     }
```

Comment:
```suggestion

    await this.throwIfCardDoesNotExistOrIsAlreadyLinked(tokenCode);

```
Nitpick, I think we do not need the if, since tokenCode is a required property